src/child.rs

```diff
@@ -169,9 +172,12 @@ impl<'s> RemoteChild<'s> {
 
         // Execute them concurrently to avoid the pipe buffer being filled up
         // and cause the remote process to block forever.
-        let (status, stdout, stderr) = try_join!(self.wait(), stdout_read, stderr_read)?;
+        let (stdout, stderr) = try_join!(stdout_read, stderr_read)?;
         Ok(Output {
-            status,
+            // Once self.wait().await is done, the connection to the multiplex
+            // server would also be dropped and the server would stop
+            // sending data to stdout/stderr.
```

Comment:
```suggestion
            // Once self.wait().await is done, the connection to the multiplex
            // server will also be dropped and the server will stop sending
            // data to stdout/stderr.
```

Reply:
I'm not sure I follow this comment. We only get to this line if both `stdout` _and_ `stderr` have already been closed (after all, we just joined them). In fact, I'm not sure why this change actually makes a difference, since the join shouldn't return until all three futures resolve. I'm guessing maybe waiting on the client has the side-effect of prematurely closing the stdout/stderr? If so, we should comment that here.

Reply:
I observe that ssh multiplex server sends the process status to openssh-rs once it received the corresponding packet from sshd.

That packet could come before all stdout/stderr data is flushed, so self.wait().await could finish before all stdout/stderr data is completely read in, closing the connection to the ssh multiplex server and cause it to drop the channel and stop flushing data.

Reply:
Ohh, this is because with `native-mux` this `wait` isn't a process wait (which doesn't _signal_ a wait), it actively starts tear-down of the channel? That seems counter-intuitive if so. In general, the expectation is that `.wait` will _just_ wait, and not _also_ trigger shutdown.

Reply:
Ah, wait, I see what you're saying. It does feel very strange that the openssh server will send a close status message if it doesn't _also_ know that the remote process has terminated (which also means it has necessarily flushed its output). This all still implies to me that `wait` has a side-effect beyond just waiting for termination.

Reply:
> This all still implies to me that wait has a side-effect beyond just waiting for termination.

Yes, that's one of the error I didn't forsee.

When I designed the api of openssh-mux-client, I didn't realize that the process status can come before all data is flushed or closing the connection to ssh multiplex server would make it drop the channel and any data that is yet to be flushed.

Reply:
But that is going to be fixed in the next-gen proxy-client, of which the ssh multiplex server would act as a proxy and all the requests/responses are going to be produced and handled in openssh-rs.

That would make multiplexing multiple `RemoteChild`s on the same connection possible, and also providing new apis like killing remote process, receiving signal used to kill the remote process, etc.

Reply:
> 

Ohh, this is because with native-mux this wait isn't a process wait (which doesn't signal a wait), it actively starts tear-down of the channel? That seems counter-intuitive if so. In general, the expectation is that .wait will just wait, and not also trigger shutdown.

Yes and that is the problem.

I can change the api of mux-client, but since `RemoteChild::wait` takes `self` by value, I guess that won't help anyway.

Reply:
I think it's counter-intuitive for `RemoteChild::wait` to _also_ trigger termination. That's not generally how `wait` works, and probably now how people expect it to work. But you are right that it's likely a breaking change. The fact that it takes `self` shouldn't make a difference though — if it blocks until the status is received, it won't drop `self` (which I assume is what triggers the termination).

In either case, I'd like this comment to reflect what's going on a bit better. Maybe (if it's correct):

>  The `self.wait()` future terminates the stdout and stderr futures when it resolves, even if there may still be more data arriving from the server. Therefore, we wait for them _first_, and only once they're complete do we wait for the process to have terminated.

Reply:
I've updated the comment.